File: matrix/matrix_multiplication.go
package main

import (
	"errors"
	"fmt"
	"reflect"
)

// Multiply performs matrix multiplication
func Multiply(A, B [][]int) ([][]int, error) {
	if len(A) == 0 || len(B) == 0 || len(A[0]) == 0 || len(B[0]) == 0 {
		return nil, errors.New("matrices cannot be empty")
	}

	rowsA := len(A)
	colsA := len(A[0])
	rowsB := len(B)
	colsB := len(B[0])

	if colsA != rowsB {
		return nil, errors.New("matrix dimensions mismatch for multiplication")
	}

	result := make([][]int, rowsA)
	for i := range result {
		result[i] = make([]int, colsB)
	}

	for i := 0; i < rowsA; i++ {
		for j := 0; j < colsB; j++ {
			for k := 0; k < colsA; k++ {
				result[i][j] += A[i][k] * B[k][j]
			}
		}
	}

	return result, nil
}

func main() {
	A := [][]int{{1, 2}, {3, 4}}
	B := [][]int{{5, 6}, {7, 8}}
	expected := [][]int{{19, 22}, {43, 50}}

	result, err := Multiply(A, B)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	if reflect.DeepEqual(result, expected) {
		fmt.Println("Test 1 Passed")
	} else {
		fmt.Println("Test 1 Failed")
	}
}

File: matrix/matrix_multiplication.rs
/**
 * Matrix Multiplication implementation
 */

pub fn multiply(a: &Vec<Vec<i32>>, b: &Vec<Vec<i32>>) -> Result<Vec<Vec<i32>>, String> {
    if a.is_empty() || b.is_empty() || a[0].is_empty() || b[0].is_empty() {
        return Err("Matrices cannot be empty".to_string());
    }

    let rows_a = a.len();
    let cols_a = a[0].len();
    let rows_b = b.len();
    let cols_b = b[0].len();

    if cols_a != rows_b {
        return Err("Matrix dimensions mismatch for multiplication".to_string());
    }

    let mut result = vec![vec![0; cols_b]; rows_a];

    for i in 0..rows_a {
        for j in 0..cols_b {
            for k in 0..cols_a {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_multiply() {
        let a = vec![vec![1, 2], vec![3, 4]];
        let b = vec![vec![5, 6], vec![7, 8]];
        let expected = vec![vec![19, 22], vec![43, 50]];
        assert_eq!(multiply(&a, &b).unwrap(), expected);

        let c = vec![vec![1, 2, 3]];
        let d = vec![vec![4], vec![5], vec![6]];
        let expected2 = vec![vec![32]];
        assert_eq!(multiply(&c, &d).unwrap(), expected2);
    }
}

fn main() {
    let a = vec![vec![1, 2], vec![3, 4]];
    let b = vec![vec![5, 6], vec![7, 8]];
    match multiply(&a, &b) {
        Ok(res) => println!("Result: {:?}", res),
        Err(e) => println!("Error: {}", e),
    }
}
